vignettes/release-announcement.Rmd

```diff
@@ -0,0 +1,131 @@
+
+```{r}
+range <- new_class("range",
+  properties = list(
+    start = class_double,
+    end = class_double
+  ),
+  validator = function(self) {
+    if (length(self@start) != 1) {
+      "@start must be length 1"
+    } else if (length(self@end) != 1) {
+      "@end must be length 1"
+    } else if (self@end < self@start) {
+      "@end must be greater than or equal to @start"
+    }
+  }
+)
+```
+
+`new_class()` returns the class object, which is also the function (aka constructor) you use to create instances of the class:
+
+```{r}
+x <- range(start = 1, end = 10)
+x
+```
+
+### Properties
+
+The data possessed by an object is called its **properties**.
+Use `@` to get and set properties:
+
+```{r}
+x@start
+
+x@end <- 20
+x
+```
+
+Properties are automatically validated against the type declared in `new_class()` (`double` in this case), and with the class **validator**:
+
+```{r, error = TRUE}
+x@end <- "x"
+
+x@end <- -1
+```
+
+### Generics and methods
+
+Like S3 and S4, S7 uses **functional OOP** where methods belong to **generic** functions, and method calls look like regular function calls: `generic(object, arg2, arg3)`.
+A generic uses the types of its arguments to automatically pick an appropriate implementation called a **method**.
+
+You can create a new generic with `new_generic()`: the first argument is the generic name (used in error messages) and the second gives the arguments used for dispatch.
+The third, and optional argument, supplies the body of the generic.
+This is only needed if your generic has additional arguments that aren't used for method dispatch.
+
+```{r}
+inside <- new_generic("inside", "x")
+```
+
+Once you have a generic, you can define a method for a specific class with `method<-`:
+
+```{r}
+# Add a method for our class
+method(inside, range) <- function(x, y) {
+  y >= x@start & y <= x@end
+}
+
+inside(x, c(0, 5, 10, 15))
+```
+
+You can use `method<-` to register methods for base types on S7 generics, and S7 classes on S3 or S4 generics.
+See [`vignette("compatibility")`](https://rconsortium.github.io/OOP-WG/articles/compatibility.html).
+for more details.
+
+Printing the generic will reveal its current methods:
+
+```{r}
+inside
+```
+
+And you can use that hint to get the method implementation:
+
+```{r}
+method(inside, range)
+```
+
+## Known limitations
+
+While we are very happy with the overall design of S7, we still haven't worked out all the details.
+Currently there are two limitations you should be aware of:
+
+-   Can't `dput()` an S7 object and you can't interactively debug methods (either with `browser()` or `debug()`). We hope to fix this in a future patch release of R.
+-   It's possible to serialize S7 object to disk (with `saveRDS()` and friends) but serialization is not yet fully thought out, and currently will currently save the entire class specification along with each object. This may change in the future.
```

Comment:
```suggestion
-   It's possible to serialize S7 object to disk (with `saveRDS()` and friends) but serialization is not yet fully thought out, and currently will save the entire class specification along with each object. This may change in the future.
```

Reply:
grammar again:  Need an article before  `S7 object`. I'd use _an_  but possibly _a_ is more correct? (Theory: It's not a vowel in writing but a vowel in *sound* / *pronunciation* that leads to the "n" ligature)

Reply:
I meant to say objects here.